relays/client-substrate/src/client/caching.rs

```diff
@@ -71,14 +74,40 @@ impl<C: Chain, B: Client<C>> CachingClient<C, B> {
 			data: Arc::new(ClientData {
 				grandpa_justifications: Arc::new(Mutex::new(None)),
 				beefy_justifications: Arc::new(Mutex::new(None)),
-				header_hash_by_number_cache: Cache::new(chain_state_capacity),
-				header_by_hash_cache: Cache::new(chain_state_capacity),
-				block_by_hash_cache: Cache::new(chain_state_capacity),
-				raw_storage_value_cache: Cache::new(1_024),
-				state_call_cache: Cache::new(1_024),
+				header_hash_by_number_cache: Arc::new(RwLock::new(Cache::new(
+					chain_state_capacity,
+				))),
+				header_by_hash_cache: Arc::new(RwLock::new(Cache::new(chain_state_capacity))),
+				block_by_hash_cache: Arc::new(RwLock::new(Cache::new(chain_state_capacity))),
+				raw_storage_value_cache: Arc::new(RwLock::new(Cache::new(1_024))),
+				state_call_cache: Arc::new(RwLock::new(Cache::new(1_024))),
 			}),
 		}
 	}
+
+	/// Try to get value from the cache, or compute and insert it using given future.
+	async fn get_or_insert_async<K: Clone + std::fmt::Debug + Eq + std::hash::Hash, V: Clone>(
+		&self,
+		cache: &Arc<RwLock<Cache<K, V>>>,
+		key: &K,
+		with: impl std::future::Future<Output = Result<V>>,
+	) -> Result<V> {
+		// try to get cached value first using read lock
+		{
+			let cache = cache.read().await;
+			if let Some(value) = cache.get(key) {
+				return Ok(value.clone())
+			}
+		}
+
+		// let's compute the value without holding any locks - it may cause additional misses and
+		// double insertions, but that's better than holding a lock for a while
+		let value = with.await?;
+
+		// insert/update the value in the cache
+		cache.write().await.insert(key.clone(), value.clone());
```

Comment:
maybe unnecesserry clone for `value.clone()` here?

Reply:
No, it is also used below (in return)

Reply:
oh, yeah, sorry :)